stan/math/rev/functor/ad_stack_matrix.hpp

```diff
@@ -0,0 +1,111 @@
+#ifndef STAN_MATH_REV_FUNCTOR_AD_STACK_MATRIX_HPP
+#define STAN_MATH_REV_FUNCTOR_AD_STACK_MATRIX_HPP
+
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/rev/core.hpp>
+
+namespace stan {
+namespace math {
+
+template <typename MatrixType>
+class AD_stack_matrix : public Eigen::Map<MatrixType> {
```

Comment:
Need docs

Reply:
I could also just go for `stack_matrix`.

Reply:
Done. Imo just stack_matrix could be mistaken as having something to do with `std::stack` or some other stack.

Reply:
What about instead of using the word `stack` we use the word `arena`? Tho' we use the word `stack` all the time

Reply:
Arena would be fine if it was used anywhere. Is anything wrong with AD_stack in name? I think it is not too long and the meaning is quite clear .

Reply:
The `AD` bothers me a bit. Like it seems like something we'd do with `CamelCase` names.

So seeing the `AD_stack_matrix` kinda makes me want `ad_stack_matrix`. Then the issue is basically everything in `stan::math` could be prefixed with `ad_` (it's an autodiff library, after all) and so that just leaves me with `stack_matrix`.

I'm happy switching to `arena` instead of `stack` to avoid confusion with whatever other stacks (the CPU stack/`std::stack` s, etc.).

In that case we'd have as a collection of things: `arena_matrix`, `to_arena`, `arena_t` and whatever else.

Reply:
OK. I renamed everything to arena_*.